File: app/src-tauri/src/windowprefs.rs
//! The window's layout and the operator's theme, **handed to the page as it is created** and
//! written back when the layout changes (`charter_core::windowprefs` holds the files and the
//! reasons).
//!
//! **Injected, never fetched.** A command is asynchronous, so a layout the window asked for
//! would land after the first paint and the window would paint the default arrangement and then
//! re-lay itself out — the flash ADR 0038 removed. So `lib.rs` builds the main window itself,
//! from its entry in `tauri.conf.json`, with [`creation_script`] as its initialization script:
//! Tauri runs that before the page's own scripts, so `main.tsx` finds both readings already in
//! the page and the first frame is drawn from them.
//!
//! **Main frame only.** `initialization_script` and not `…_for_all_frames`: an extension's panel
//! is a frame of its own, and the operator's arrangement is nothing it needs to be handed.

use std::path::Path;

use charter_core::windowprefs::{self, Reading};

/// The global the page reads both readings from. `app/src/windowprefs.ts` names it too; the
/// test below holds the script to it, and `windowprefs.test.ts` holds the page to it.
pub(crate) const GLOBAL: &str = "__CHARTER_AT_CREATION__";

/// Both readings, as the page receives them.
#[derive(Debug, serde::Serialize)]
struct AtCreation {
    layout: Reading,
    theme: Reading,
}

/// The script that puts both readings on the page before any of the page's own code runs.
///
/// **What is read from disk reaches the page as JSON data, never as code.** The readings are
/// serialised by `serde_json`, which escapes every quote and control character in a string, so
/// nothing in a hand-edited file can end the literal it sits in; and the value is frozen, so
/// nothing that runs later can quietly rewrite what the launch read.
///
/// A machine with no config home hands the page two empty readings: the default arrangement and
/// the built-in theme, which is what such a machine has always had.
pub(crate) fn creation_script(config_root: Option<&Path>) -> String {
    let at = match config_root {
        Some(root) => AtCreation {
            layout: windowprefs::read_layout(root),
            theme: windowprefs::read_theme(root),
        },
        None => AtCreation {
            layout: Reading::default(),
            theme: Reading::default(),
        },
    };
    let json = serde_json::to_string(&at).expect("a reading is plain data serde can always write");
    format!(
        "Object.defineProperty(window, {GLOBAL:?}, {{ value: Object.freeze({json}), \
         writable: false, configurable: false }});"
    )
}

/// Keeps the window's layout, replacing what the file held.
///
/// `text` is the document as JSON text; the core parses it and writes it back out itself, so
/// what lands on disk is always a document charter wrote.
#[tauri::command]
#[specta::specta]
pub async fn write_layout(text: String) -> Result<(), String> {
    let root = config_root()?;
    tauri::async_runtime::spawn_blocking(move || windowprefs::write_layout(&root, &text))
        .await
        .map_err(|err| format!("keeping the layout did not finish: {err}"))?
        .map_err(|err| err.to_string())
}

/// Moves the arrangement web storage held into the layout file — **only when there is no file
/// yet**. Answers whether it did, so the window knows the old key can go.
#[tauri::command]
#[specta::specta]
pub async fn adopt_layout(text: String) -> Result<bool, String> {
    let root = config_root()?;
    tauri::async_runtime::spawn_blocking(move || windowprefs::adopt_layout(&root, &text))
        .await
        .map_err(|err| format!("moving the layout did not finish: {err}"))?
        .map_err(|err| err.to_string())
}

/// The config home, or the reason charter keeps no layout on this machine.
fn config_root() -> Result<std::path::PathBuf, String> {
    charter_core::machine::config_root().ok_or_else(|| {
        "this machine has no config home, so charter cannot keep the layout".to_owned()
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The object the script defines, read back the way a page would get it.
    fn defined(script: &str) -> serde_json::Value {
        let start = script.find("Object.freeze(").expect("frozen") + "Object.freeze(".len();
        let end = script.rfind("), writable").expect("closed");
        serde_json::from_str(&script[start..end]).expect("the value is JSON")
    }

    #[test]
    fn the_page_is_handed_the_layout_and_the_theme_on_disk() {
        let home = tempfile::tempdir().unwrap();
        let dir = charter_core::machine::dir(home.path());
        std::fs::create_dir_all(&dir).unwrap();
        std::fs::write(
            dir.join(windowprefs::LAYOUT),
            r#"{"version":1,"regions":[{"id":"aside","side":"left","order":0}]}"#,
        )
        .unwrap();
        std::fs::write(dir.join(windowprefs::THEME), r#"{"name":"Mine"}"#).unwrap();

        let script = creation_script(Some(home.path()));
        assert!(script.starts_with(&format!("Object.defineProperty(window, \"{GLOBAL}\"")));
        let at = defined(&script);
        assert_eq!(at["layout"]["found"], true);
        assert_eq!(at["layout"]["document"]["regions"][0]["side"], "left");
        assert_eq!(at["theme"]["document"]["name"], "Mine");
    }

    #[test]
    fn a_hand_edited_file_reaches_the_page_as_data_and_never_as_code() {
        // Whatever the operator — or anybody who can write the file — puts in a string, it is a
        // string when it arrives.
        let home = tempfile::tempdir().unwrap();
        let dir = charter_core::machine::dir(home.path());
        std::fs::create_dir_all(&dir).unwrap();
        let hostile = r#"\"}); alert(1); ({\" </script>"#;
        std::fs::write(
            dir.join(windowprefs::THEME),
            format!(r#"{{"name":"{hostile}"}}"#),
        )
        .unwrap();
        let at = defined(&creation_script(Some(home.path())));
        assert_eq!(
            at["theme"]["document"]["name"],
            "\"}); alert(1); ({\"\u{2028}</script>"
        );
    }

    #[test]
    fn a_broken_layout_reaches_the_page_as_a_reason_and_no_document() {
        let home = tempfile::tempdir().unwrap();
        let dir = charter_core::machine::dir(home.path());
        std::fs::create_dir_all(&dir).unwrap();
        std::fs::write(dir.join(windowprefs::LAYOUT), "{ nope").unwrap();
        let at = defined(&creation_script(Some(home.path())));
        assert_eq!(at["layout"]["found"], true);
        assert_eq!(at["layout"]["document"], serde_json::Value::Null);
        assert!(
            at["layout"]["trouble"]
                .as_str()
                .is_some_and(|why| why.contains("is not JSON"))
        );
    }

    #[test]
    fn a_machine_with_no_config_home_is_handed_nothing_to_draw_from() {
        let at = defined(&creation_script(None));
        assert_eq!(at["layout"]["found"], false);
        assert_eq!(at["theme"]["found"], false);
    }
}

File: app/src/windowprefs.ts
import { useSyncExternalStore } from "react";
import type { AlertRow } from "./bindings";
import { load, type Theme } from "./theme/theme";

/**
 * **What the window was handed as it was created**: the operator's layout file and theme file,
 * read by the Rust side before the window existed (`src-tauri/src/windowprefs.rs`).
 *
 * Both are files beside `machine.json` (`docs/design-system.md` has their formats), and both
 * are needed by the very first frame — a layout drawn after it is a re-layout the operator
 * watches happen. A Tauri command is asynchronous, so they are not fetched: the window's
 * initialization script defines {@link GLOBAL} before any of this code runs, and reading it is
 * a property access.
 */

/** The global the initialization script defines. `windowprefs.rs` names it too. */
export const GLOBAL = "__CHARTER_AT_CREATION__";

/** One file's reading — `charter_core::windowprefs::Reading`, as serialised. */
export type Reading = {
  /** Where the file is, or would be. */
  path: string;
  /** Whether there is a file at all. */
  found: boolean;
  /** What it holds, when charter can use it. Untrusted: a file anybody may have edited. */
  document: unknown;
  /** Why a file that is there is not in force. */
  trouble: string | null;
};

/** Both readings. */
export type AtCreation = { layout: Reading; theme: Reading };

const NOTHING: Reading = { path: "", found: false, document: null, trouble: null };

/**
 * What the window was handed, or two empty readings when it was handed nothing — a page opened
 * outside the app, or a test that has not set one. Checked field by field rather than trusted,
 * because a reading the window cannot make sense of must cost the preference and never the
 * window.
 */
export function atCreation(): AtCreation {
  const given = (globalThis as Record<string, unknown>)[GLOBAL];
  const held =
    given !== null && typeof given === "object" ? (given as Record<string, unknown>) : {};
  return { layout: reading(held.layout), theme: reading(held.theme) };
}

function reading(raw: unknown): Reading {
  if (raw === null || typeof raw !== "object") return NOTHING;
  const from = raw as Record<string, unknown>;
  return {
    path: typeof from.path === "string" ? from.path : "",
    found: from.found === true,
    document: from.document ?? null,
    trouble: typeof from.trouble === "string" ? from.trouble : null,
  };
}

/**
 * **What charter says about this machine rather than about a project** — a layout file it could
 * not use, a theme file with a token it had to put right, a layout it could not keep.
 *
 * The alerts drawer is where charter says what is wrong, and every alert it had was about a
 * plane. These are not: they are about files in the operator's config directory, and they are
 * known to the window rather than to the core, because it is the window's vocabulary — the
 * regions and the tokens — that decides whether a file said something usable. So the window
 * keeps them here, one per subject, and the drawer draws them above the projects.
 */
const said = new Map<string, AlertRow>();
let listed: AlertRow[] = [];
const listeners = new Set<() => void>();

/** Says something about this machine under `subject`, replacing what was said under it before,
 *  or takes it back with `undefined`. */
export function sayAboutThisMachine(subject: string, row: Omit<AlertRow, "subject"> | undefined) {
  if (row === undefined) {
    if (!said.delete(subject)) return;
  } else {
    said.set(subject, { ...row, subject });
  }
  listed = [...said.values()];
  for (const listener of listeners) listener();
}

/** What is being said about this machine right now. */
export function aboutThisMachine(): AlertRow[] {
  return listed;
}

/** {@link aboutThisMachine}, for a component that redraws when it changes. */
export function useAboutThisMachine(): AlertRow[] {
  return useSyncExternalStore((listener) => {
    listeners.add(listener);
    return () => listeners.delete(listener);
  }, aboutThisMachine);
}

/** {@link theirTheme} of what the window was handed, read once: so drawing it again after a
 *  project's own theme is the same object, and repaints nothing that is already drawn. */
let theirs: { theme: Theme | undefined } | undefined;

export function theirThemeOnce(): Theme | undefined {
  theirs ??= { theme: theirTheme() };
  return theirs.theme;
}

/** For tests: read the handed theme again. */
export function forgetTheirTheme() {
  theirs = undefined;
}

/**
 * **The operator's own theme, or none** (M6.7) — `charter/theme.json`, the address
 * `docs/design-system.md` gave it, and until now read by nothing.
 *
 * It arrives with the window, so `main.tsx` draws it **before the first frame**: an operator
 * with a theme of their own never sees the built-in painted first. `theme.load` is what judges
 * it — every token against the vocabulary and the hex grammar — and whatever it had to put
 * right is said in the alerts drawer, with the file's path, because a colour that silently
 * came out as the built-in's is a theme that looks like it was ignored.
 *
 * A file that is not a theme at all is said the same way, and the window keeps the built-in.
 */
export function theirTheme(reading: Reading = atCreation().theme): Theme | undefined {
  const where = reading.path || "the theme file";
  if (reading.trouble !== null) {
    sayAboutThisMachine("theme", {
      severity: "warn",
      detail: `${reading.trouble} — the window is drawn in the built-in theme`,
      remedy: `fix ${where}, or delete it to keep the built-in`,
    });
    return undefined;
  }
  if (reading.document === null) return undefined;
  const { theme, complaints } = load(reading.document);
  if (complaints.length > 0) {
    sayAboutThisMachine("theme", {
      severity: "warn",
      detail: `${where}: ${complaints.map((one) => one.said).join("; ")}`,
      remedy: `fix ${where}; everything else in it is drawn`,
    });
  }
  return theme;
}
